Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,667 @@
+    /// The request ID, which identifies the request that triggered the function invocation.
+    public var requestID: String { get }
+
+    /// The AWS X-Ray tracing header.
+    public var traceID: String { get }
+
+    /// The ARN of the Lambda function, version, or alias that's specified in the invocation.
+    public var invokedFunctionARN: String { get }
+
+    /// The timestamp that the function times out.
+    public var deadline: DispatchWallTime { get }
+
+    /// For invocations from the AWS Mobile SDK, data about the Amazon Cognito identity provider.
+    public var cognitoIdentity: String? { get }
+
+    /// For invocations from the AWS Mobile SDK, data about the client application and device.
+    public var clientContext: String? { get }
+
+    /// `Logger` to log with.
+    ///
+    /// - note: The `LogLevel` can be configured using the `LOG_LEVEL` environment variable.
+    public var logger: Logger { get }
+
+    /// Schedule background work in lambda. This work must be completed before
+    /// the runtime can ask for further events. Note that AWS will continue to
+    /// charge you after the response has been returned, but work is still
+    /// processing in the background.
+    public func addBackgroundTask(_ body: sending @escaping () async -> ())
+}
+```
+
+### StreamingLambdaHandler
+
+The new `StreamingLambdaHandler` protocol is the base protocol to implement a Lambda function. Most users will not use
+this protocol and instead use the `LambdaHandler` protocol defined in the **Codable Support** section.
+
+```swift
+/// The base StreamingLambdaHandler protocol
+public protocol StreamingLambdaHandler: ~Copyable {
+    /// The business logic of the Lambda function
+    /// - Parameters:
+    ///   - event: The invocation's input data
+    ///   - responseWriter: A ``LambdaResponseWriter`` to write the invocations response to.
+    ///                     If no response or error is written to the `responseWriter` it will
+    ///                     report an error to the invoker.
+    ///   - context: The LambdaContext containing the invocation's metadata
+    mutating func handle(_ event: ByteBuffer, responseWriter: consuming some LambdaResponseWriter, context: LambdaContext) async
+}
+```
+
+The protocol is defined in a way that supports a broad range of use-cases. The handle method is marked as `mutating` to
+allow handlers to be implemented with a `struct`. Users can opt to even mark the handler as non-copyable.
+
+An implementation that sends the number 1 to 10 every 500ms could look like this:
+
+```swift
+struct SendNumbersWithPause: StreamingLambdaHandler {
+    func handle(
+        _ event: ByteBuffer,
+        responseWriter: consuming some LambdaResponseWriter,
+        context: LambdaContext
+    ) async {
+        for i in 1...10 {
+            responseWriter.write(ByteBuffer(string: #"\#(i)\n\r"#))
+            try? await Task.sleep(for: .milliseconds(500))
+        }
+        responseWriter.finish()
```

Comment:
What happens if the user doesn't call `finish` explicitly?

Reply:
So it is safe to call `finish()` multiple times?

Reply:
`finish` is consuming so you can't call it multiple times